napari/utils/key_bindings.py

```diff
@@ -182,6 +186,284 @@ def normalize_key_combo(key_combo):
+        # shortcuts may have been bound before the command was actually
+        # registered, remove it from orphan and bind shortcuts.
+        if command in self._orphans:
+            # print('Found orphan shortcut')
+            sht = self._orphans.pop(command)
+            self.bind_shortcut(name, sht)
+        self._update_gui_elements(name)
+
+    def _update_gui_elements(self, name):
+        """
+        Update the description and shortcuts of all the (known) gui elements.
+        """
+        if name not in self._actions:
+            return
+        buttons = self._buttons.get(name, [])
+        desc = self._actions[name].description
+
+        # update buttons with shortcut and description
+        if name in self._shortcuts:
+            sht = self._shortcuts[name]
+            sht_str = f' ({format_shortcut(sht)})'
+        else:
+            sht = ''
+            sht_str = ''
+
+        for button in buttons:
+            button.setToolTip(desc + sht_str)
+        if name in self._qactions:
+            action = self._actions[name]
+            qaction = self._qactions[name]
+            # We can't set the shortcut yet, or this will later break the layers
+            # when we use the action manager for those. Setting the key sequence
+            # there will _also_ bind the shortcuts but at the window level which
+            # has a higher priority. SO this conflict with KeymapProvider
+            #
+            # qaction.setShortcut(
+            #     QKeySequence(sht.replace('-', '+').replace('Control', 'Ctrl'))
+            # )
+            menu_name = name
+            if ' ' not in menu_name:
+                components = [
+                    word.capitalize() for word in menu_name.split('_')
+                ]
+                menu_name = ' '.join(components)
+            qaction.setText(menu_name)
+            qaction.setStatusTip(action.description)
+
+    def _update_shortcut_bindings(self, name):
+        """
+        Update the key mappable for given action name
+        to trigger the action within the given context and
+        make all the corresponding UI element flash if possible.
+        """
+        if name not in self._actions:
+            return
+        action = self._actions[name]
+        if name not in self._shortcuts:
+            return
+        sht = self._shortcuts.get(name)
+        keymappable = action.keymappable
+        if hasattr(keymappable, 'bind_key'):
+
+            def flash(*args):
+
+                # here we do not need call with context as we are still using
+                # KeymapProvider API that will pass their own instance as the
+                # first argument. This will likely need to be rethought in the
+                # future.
+                action.command(*args)
+                for b in self._buttons.get(name, []):
+                    b._animation.start()
+
+            keymappable.bind_key(sht)(flash)
+
+    def bind_button(self, name, button):
```

Comment:
See comment on using toolbars and adding actions to the toolbars.

https://github.com/napari/napari/pull/2495/files#r605249488